src/tsolvers/egraph/EgraphSolver.C

```diff
@@ -221,6 +221,34 @@ PTRef Egraph::getSuggestion( )
   return PTRef_Undef;
 }
 
+lbool Egraph::getPolaritySuggestion(PTRef p)
+{
+    if (!isInformed(p)) { return l_Undef; }
+    // MB: it could be uninterpreted predicate! No suggestion in that case
+    if (!logic.isEquality(p) && !logic.isDisequality(p)) { return l_Undef; }
+    bool equality = logic.isEquality(p);
+    const Pterm& term = logic.getPterm(p);
+    if(term.size() > 2) { return l_Undef; } // For now focus on 2 arguments
+    PTRef lhs = term[0];
+    PTRef rhs = term[1];
+    assert(enode_store.termToERef.has(lhs) && enode_store.termToERef.has(rhs));
+    if (!enode_store.termToERef.has(lhs) || !enode_store.termToERef.has(rhs)) { return l_Undef; }
+    ERef e_lhs = termToERef(lhs);
```

Comment:
Looks like the code 236 -- 239 is the same as the beginning of Egraph::unmergeable.  I think the same functionality would be available with less code by just calling bool res = unmergeable(e_lhs, e_rhs, tmp);

Reply:
The change you suggest would result in a different behaviour.
With the current implementation, EGraph suggests to make things equal only if they ARE ALREADY in the same equivalence classes. (And suggests to make them not equal if they are unmergeable).

In your scenario, Egraph would suggest to make things equal if it CANNOT prove that they are unmergeable. So it would make suggestions more often.

My original idea was for the Egraph would only suggest polarity if it KNOWS that the opposite polarity would lead to conflict. But I don't mind making it more aggressive suggestions to make things equal. If you think it's better that way I will change it.

Reply:
I think you're right, just keep it the way you first coded it.